src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+             * Idea is 's/[0-9]+//' */
+            channel_pure_name = calloc(1, channel_name_len + 1);
+            original_channel_pure_name = channel_pure_name;
+            for (i = 0; i < channel_name_len; i++) {
+                if (isalpha(channel->name[i]) || channel->name[i] == '-' || channel->name[i] == '_')
+                    *channel_pure_name++ = channel->name[i];
+                else if (isdigit(channel->name[i])) {
+                    modified = true;
+                    continue;
+                }
+            }
+            if (modified)
+                return original_channel_pure_name;
+            else
+                free(original_channel_pure_name);
+        }
+    }
+
+    return NULL;
+}
+
+static void
+channel_get_scale(struct sol_iio_channel *channel)
+{
+    char path[PATH_MAX + 1];
+    int len;
+    char *channel_pure_name = NULL;
+    struct sol_iio_device *device = channel->device;
+
+#define GET_SCALE(_name) \
+    do { \
+        len = snprintf(path, sizeof(path), CHANNEL_SCALE_PATH, device->device_id, \
+            _name); \
+        if (len > 0 && sol_util_read_file(path, "%lf", &channel->scale) > 0) { \
+            goto end; \
+        } \
+    } while (0)
+
+    GET_SCALE(channel->name);
+    /* No scale. If channel has x, y, z or Y component, look for scale file without it */
+    channel_pure_name = channel_get_pure_name(channel);
+    GET_SCALE(channel_pure_name);
+
+    SOL_INF("Could not get scale for channel [%s] in device%d. Assuming 1.0",
+        channel->name, device->device_id);
+    channel->scale = 1.0;
+
+#undef GET_SCALE
+
+end:
+    free(channel_pure_name);
+}
+
+static void
+channel_get_offset(struct sol_iio_channel *channel)
+{
+    char path[PATH_MAX + 1];
+    int len;
+    char *channel_pure_name = NULL;
+    struct sol_iio_device *device = channel->device;
+
+#define GET_OFFSET(_name) \
+    do { \
+        len = snprintf(path, sizeof(path), CHANNEL_OFFSET_PATH, device->device_id, \
+            _name); \
+        if (len > 0 && sol_util_read_file(path, "%d", &channel->offset) > 0) { \
+            goto end; \
+        } \
+    } while (0)
+
+    GET_OFFSET(channel->name);
+    /* No scale. If channel has x, y or z component, look for scale file without it */
+    channel_pure_name = channel_get_pure_name(channel);
+    GET_OFFSET(channel_pure_name);
+
+    SOL_INF("Could not get offset for channel [%s] in device%d. Assuming 0",
+        channel->name, device->device_id);
+    channel->offset = 0;
+
+#undef GET_OFFSET
+
+end:
+    free(channel_pure_name);
+}
+
+static void
+iio_del_channel(struct sol_iio_channel *channel)
+{
+    free(channel->name);
+
+    free(channel);
+}
+
+void
+sol_iio_close(struct sol_iio_device *device)
+{
+    int i;
+    struct sol_iio_channel *channel;
+
+    SOL_PTR_VECTOR_FOREACH_REVERSE_IDX (&device->channels, channel, i) {
+        iio_del_channel(channel);
+        sol_ptr_vector_del(&device->channels, i);
+    }
```

Comment:
it's better to loop the vector and delete contents, then free the array at once with `sol_ptr_vector_clear()` than to delete one vector element at time, which will trigger lots of `realloc()`